File: 1209-remove-all-adjacent-duplicates-in-string-ii/1209-remove-all-adjacent-duplicates-in-string-ii.cpp
class Solution {
public:
    string removeDuplicates(string s, int k) {
        stack<pair<char, int>> st;
        
        st.push({s[0], 1});
        
        for(int i=1; i<s.size(); i++){
            if(!st.empty()){
                if(st.top().first == s[i]){
                    pair<char, int> topPair = st.top();
                    st.push({s[i], topPair.second + 1});
                } else{
                    st.push({s[i], 1});
                }
            } else{
                st.push({s[i], 1});
            }
            
            if(st.top().second == k){
                int count = 0;
                
                while(!st.empty() && count<k){
                    st.pop();
                    count++;
                }
            }
        }
        
        string ans = "";

        while(!st.empty()){
            ans += st.top().first;
            st.pop();
        }
        reverse(begin(ans), end(ans));
        
        return ans;
    }
};

File: 1209-remove-all-adjacent-duplicates-in-string-ii/1209-remove-all-adjacent-duplicates-in-string-ii.java
class Solution {
    static class Pair {
        char first;
        int second;
        
        Pair(char first, int second) {
            this.first = first;
            this.second = second;
        }
    }
    public String removeDuplicates(String s, int k) {
        ArrayDeque<Pair> st = new ArrayDeque<>();
        
        st.push(new Pair(s.charAt(0), 1));
        
        for (int i = 1; i < s.length(); i++) {
            if (!st.isEmpty()) {
                if (st.peek().first == s.charAt(i)) {
                    Pair topPair = st.peek();
                    st.push(new Pair(s.charAt(i), topPair.second + 1));
                } else {
                    st.push(new Pair(s.charAt(i), 1));
                }
            } else {
                st.push(new Pair(s.charAt(i), 1));
            }
            
            if (st.peek().second == k) {
                int count = 0;
                
                while (!st.isEmpty() && count < k) {
                    st.pop();
                    count++;
                }
            }
        }
        
        StringBuilder ans = new StringBuilder();
        
        while (!st.isEmpty()) {
            ans.append(st.peek().first);
            st.pop();
        }
        
        return ans.reverse().toString();
    }
}